Add SPDM 1.3 feature GET_ENDPOINT_INFO

Li-Aaron opened the pull request:
Fix the issue: #2294

ENDPOINT_INFO is SPDM spec 1.3 new feature.
Add ENDPOINT_INFO for Libspdm.

Note:
Encapsulate ENDPOINT_INFO and document update will be handled in separate patches.

Li-Aaron commented on library/spdm_requester_lib/libspdm_req_get_endpoint_info.c:
```diff
@@ -0,0 +1,348 @@
+         * get_measurement have this check
+         * measurement: this field shall contain the slot number of the certificate chain specified in the GET_MEASUREMENTS request
+         * endpoint_info: SlotID that identifies the certificate chain whose leaf certificate is used to sign the response
+         */
```
Is the spec indicating responder's `slot_id` should be same with responder?
> **get_measurement** have the check that `slot_id` should be same between requester and responder (which is required by spec)

In Spec:
> **measurement**: this field shall contain the slot number of the certificate chain specified in the GET_MEASUREMENTS request
> **endpoint_info**: SlotID that identifies the certificate chain whose leaf certificate is used to sign the response

Li-Aaron commented on library/spdm_requester_lib/libspdm_req_get_endpoint_info.c:
```diff
@@ -0,0 +1,342 @@
+
+    /* -=[Process Response Phase]=- */
+    if ((request_attributes & SPDM_GET_ENDPOINT_INFO_REQUEST_ATTRIBUTE_SIGNATURE_REQUESTED) != 0) {
+        /* Q: Does the spec implicitly indicate that the responder's slot_id should be the same as the responder? */
```
In Spec:

**(496) MEASUREMENTS** `Param2`: Bit [3:0]. Shall be the SlotID . If this message contains a signature, this field shall contain the slot number of the certificate chain specified in the
GET_MEASUREMENTS request
**(977) ENDPOINT_INFO** `Param2`: Bit [3:0]. SlotID that identifies the certificate chain whose leaf certificate is used to sign the response.

steven-bellock replied:
`ENDPOINT_INFO.Param2` is equal to `GET_ENDPOINT_INFO.Param2`.

Li-Aaron replied:
Got it, thanks.

steven-bellock commented on include/library/spdm_lib_config.h:
```diff
@@ -82,6 +82,10 @@
 #define LIBSPDM_ENABLE_CAPABILITY_SET_KEY_PAIR_INFO_CAP 1
 #endif
 
+#ifndef LIBSPDM_ENABLE_ENDPOINT_INFO_CAP
```
There also needs to be a `LIBSPDM_ENDPOINT_INFO_SUPPORT` macro, to enable this for endpoints that want to send `GET_ENDPOINT_INFO`. See the difference between `LIBSPDM_ENABLE_CAPABILITY_EVENT_CAP` and `LIBSPDM_EVENT_RECIPIENT_SUPPORT`.

Li-Aaron replied:
If I understand correctly, `LIBSPDM_EVENT_RECIPIENT_SUPPORT` is a macro that indicate the requester have the ability to response the encap `SEND_EVENT`.

For "endpoints that want to send `GET_ENDPOINT_INFO`", do you mean requester have the ability to respond encap `GET_ENDPOINT_INFO`?

steven-bellock replied:
No, just to send it. For example https://github.com/DMTF/libspdm/pull/3000/files#diff-6c0c726de3f2aaccd4e39d09f5bdd2c2a04be1e3776c7fd33ecba1496f36820cR8-R10 should be `LIBSPDM_ENDPOINT_INFO_SUPPORT`.

jyao1 commented on include/library/spdm_common_lib.h:
```diff
@@ -958,6 +958,31 @@ uint32_t libspdm_module_version(void);
+ * @retval LIBSPDM_STATUS_SUCCESS Success
+ * @retval LIBSPDM_STATUS_INVALID_PARAMETER Some parameters invalid or NULL
+ **/
+typedef libspdm_return_t (*libspdm_endpoint_info_device_func)(
```
Can we move this to a HAL lib, e.g. https://github.com/DMTF/libspdm/blob/main/include/hal/library/responder/measlib.h?

Li-Aaron commented:
Changes:

1. Add check for `ENDPOINT_INFO.Param2` is equal to `GET_ENDPOINT_INFO.Param2` in `GET_ENDPOINT_INFO`.
2. Add `LIBSPDM_ENDPOINT_INFO_SUPPORT` marco for endpoints that want to send `GET_ENDPOINT_INFO`.
3. Move callback `libspdm_endpoint_info_device_func` to HAL function `libspdm_generate_device_endpoint_info`.
4. Modify `LIBSPDM_ENABLE_ENDPOINT_INFO_CAP` to `LIBSPDM_ENABLE_CAPABILITY_ENDPOINT_INFO_CAP` to align with other marcos.
5. Add `LIBSPDM_ENABLE_CAPABILITY_ENDPOINT_INFO_CAP` `LIBSPDM_SEND_GET_ENDPOINT_INFO_SUPPORT` in CI configuration.

steven-bellock commented on include/library/spdm_responder_lib.h:
```diff
@@ -1,6 +1,6 @@
 /**
  *  Copyright Notice:
- *  Copyright 2021-2024 DMTF. All rights reserved.
+ *  Copyright 2021-2025 DMTF. All rights reserved.
```
This can be reverted.

steven-bellock commented on library/spdm_responder_lib/libspdm_rsp_endpoint_info.c:
```diff
@@ -0,0 +1,261 @@
+
+    /* -=[Check Parameters Phase]=- */
+    LIBSPDM_ASSERT(spdm_request->header.request_response_code == SPDM_GET_ENDPOINT_INFO);
+
```
Need to check that the negotiated SPDM version is `>= 1.3`.

steven-bellock commented on library/spdm_requester_lib/libspdm_req_get_endpoint_info.c:
```diff
@@ -0,0 +1,347 @@
+    LIBSPDM_ASSERT((slot_id != 0xF) ||
+                   (spdm_context->local_context.peer_public_key_provision_size != 0));
+
+    /* -=[Verify State Phase]=- */
```
Need to check that the negotiated SPDM version is `>= 1.3`.

steven-bellock commented on include/library/spdm_lib_config.h:
```diff
@@ -213,6 +225,13 @@
+/* This value specifies the maximum size, in bytes, of a endpoint info that can be stored in a
+ * libspdm context.
+ */
+#ifndef LIBSPDM_MAX_ENDPOINT_INFO_LENGTH
```
This is conditional on `LIBSPDM_RECORD_TRANSCRIPT_DATA_SUPPORT == 1`.

Li-Aaron commented on include/hal/library/responder/endpointinfolib.h:
```diff
@@ -0,0 +1,37 @@
+ **/
+extern libspdm_return_t libspdm_generate_device_endpoint_info(
+    uint8_t sub_code,
+    uint32_t *endpoint_info_size,
```
ok, will add `spdm_context` as well.

Li-Aaron commented on include/hal/library/responder/endpointinfolib.h:
```diff
@@ -0,0 +1,37 @@
+ *                              If not NULL, copy the device class identifier to the buffer.
+ *
+ * @retval LIBSPDM_STATUS_SUCCESS Success
+ * @retval LIBSPDM_STATUS_INVALID_PARAMETER Some parameters invalid or NULL
```
ok, will remove

Li-Aaron commented:
[Changes between bd6ee72 and a157d77](https://github.com/DMTF/libspdm/compare/bd6ee72c3dda8b1ddd5073442b25c9604afff3ad..a157d77bd114b89a7bd2148824279cdda75ae106)

1. move `include/hal/library/responder/endpointinfolib.h` to `include/hal/library/endpointinfolib.h`.
2. add `spdm_context` and `request_attribute` to HAL func `libspdm_generate_device_endpoint_info`, and change the return type from `libspdm_return_t` to `bool`, 
3. remove the `INVALID_PARAMETER` check in HAL func `libspdm_generate_device_endpoint_info`.
4. move marco `LIBSPDM_MAX_ENDPOINT_INFO_LENGTH` in `LIBSPDM_RECORD_TRANSCRIPT_DATA_SUPPORT == 1`
5. add version 1.3 check
6. move error test to error_test folder.

steven-bellock commented on include/hal/library/endpointinfolib.h:
```diff
@@ -0,0 +1,42 @@
+ *  License: BSD 3-Clause License. For full text see link: https://github.com/DMTF/libspdm/blob/main/LICENSE.md
+ **/
+
+#ifndef RESPONDER_ENDPOINTINFOLIB_H
```
`RESPONDER_ENDPOINTINFOLIB_H` -> `ENDPOINTINFOLIB_H`

steven-bellock commented on include/hal/library/endpointinfolib.h:
```diff
@@ -0,0 +1,42 @@
+ * @retval true   Success.
+ * @retval false  An error occurred when processing the device endpoint info.
+ **/
+extern bool libspdm_generate_device_endpoint_info(
```
`bool` -> `libspdm_return_t`
When libspdm 4.0 comes out all these `bool`s will become `libspdm_return_t`.

steven-bellock commented on include/hal/library/endpointinfolib.h:
```diff
@@ -0,0 +1,42 @@
+
+#endif /* LIBSPDM_ENABLE_CAPABILITY_ENDPOINT_INFO_CAP */
+
+#endif /* RESPONDER_ENDPOINTINFOLIB_H */
```
`RESPONDER_ENDPOINTINFOLIB_H` -> `ENDPOINTINFOLIB_H`

steven-bellock commented on include/internal/libspdm_common_lib.h:
```diff
@@ -335,10 +366,14 @@ typedef struct {
     libspdm_message_c_managed_buffer_t message_mut_c;
     libspdm_message_m_managed_buffer_t message_m;
+    libspdm_message_e_managed_buffer_t message_e;
+    libspdm_message_e_managed_buffer_t message_mut_e;
```
`_mut_` was used for mutual authentication, however `GET_ENDPOINT_INFO` cannot be issued during mutual authentication. Maybe `_encap_` is better suited? @jyao1 to comment.

jyao1 replied:
agree.

Li-Aaron commented:
[Changes between a157d77 and 5c60935](https://github.com/DMTF/libspdm/compare/a157d77bd114b89a7bd2148824279cdda75ae106..5c60935e7dd0329c0695b09c08e5986d8f6be712)

1. `RESPONDER_ENDPOINTINFOLIB_H` -> `ENDPOINTINFOLIB_H`
2. change the return type from `bool` to `libspdm_return_t` for HAL function `libspdm_generate_device_endpoint_info`
3. `_mut_e` -> `_encap_e`, `_mut_il1il2` -> `_encap_il1il2`.

Li-Aaron commented:
[Changes between 9176e94 and 58328a3](https://github.com/DMTF/libspdm/compare/9176e945694708015a8d63b90e555590d17fbc77..58328a3920037425d5fe599388bd3f81b8a46b92)
fix 3 error case for endpoint info response.

Li-Aaron commented:
Hi @steven-bellock ,

Thank you for your previous review! I’ve addressed your feedback and made some updates to the PR. 
Could you take another look? 
Let me know if anything else needs attention. Thanks again for your time and help!

jyao1 commented on library/spdm_requester_lib/libspdm_req_get_endpoint_info.c:
```diff
@@ -0,0 +1,351 @@
+    }
+
+
+    *ep_info_len = ep_info_data_len;
```
You cannot assign the value directly.

What if (*ep_info_len < ep_info_data_len)?

Li-Aaron replied:
[Changes between 58328a3 and 3ab7127](https://github.com/DMTF/libspdm/compare/58328a3920037425d5fe599388bd3f81b8a46b92..3ab7127c2ac4b4e59198e17222a27706d7fd5aaa)
1. add `ep_info_len` check
2. initialize `ep_info_length` before use in test cases
3. add 1 err test for `ep_info_length` too small
